### PR TITLE
Make dynamic api downstream of dynamo stateless

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96012
* #96338
* __->__ #96678

Rm unused

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1059,17 +1059,20 @@ def wrap_to_fake_tensor_and_record(
     ):
         static_shapes, reason = tensor_always_has_static_shape(e, source, is_tensor)
 
+        dynamic_indices = None
+        if hasattr(e, "_dynamo_dynamic_indices"):
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:1063]: This part is still stateful, and read off the tensor. Perhaps there is a way to make it stateless here too?

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/96699

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1409,7 +1424,7 @@ def duck_int(self, val):
     # some equality guards are nontrivial!  It would be nice to get simplified
     # output to print them too).  It's private because it's not
     # intended for normal use
-    def produce_guards(self, placeholders, sources,
+    def produce_guards(self, placeholders, sources, dynamic_indices: Set[int],
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1427]: Type annotation here is wrong

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1530,7 +1545,7 @@ def _verify(expr, potential_expr):
             for i, ss in enumerate(t.size()):
                 property_source = TensorPropertySource(source, TensorProperty.SIZE, i)
                 track_symint(property_source, ss)
-                if _is_dim_dynamic(t, i):
+                if dyn_dims and i in dyn_dims:
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1548]: this is annoying, can you force the user to always pass in a set (could be empty)

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -1340,13 +1354,14 @@ def create_unbacked_symint(self):
     # This is guaranteed to return a symbol or its negation is a sympy.Symbol,
     # but there may be a replacement that allows it to be immediately
     # simplified
-    def create_symbol(self, val: int, source: Source, dyn=False) -> "sympy.Expr":
+    def create_symbol(self, val: int, source: Source, dim_state=DIM_DYNAMISM_STATE) -> "sympy.Expr":
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1357]: As is, you need to assert dim_state is not static. But I still think it is profitable to make this handle static too.

**Pull request closed without merging**